messages/chainlink/message_handler.go

```diff
@@ -0,0 +1,369 @@
+) (messages.MessageHandlerFactory, error) {
+	config, err := ParseConfig(messageProtocolConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return &factory{
+		logger: logger,
+		config: config,
+	}, nil
+}
+
+func (f *factory) NewMessageHandler(unsignedMessage *warp.UnsignedMessage) (messages.MessageHandler, error) {
+	return &ChainlinkMessageHandler{
+		logger:                  f.logger,
+		unsignedMessage:         unsignedMessage,
+		destinationBlockchainID: f.config.DestinationBlockchainID,
+		maxFilterAdresses:       f.config.MaxFilterAdresses,
+		aggregatorsToReplicas:   f.config.AggregatorsToReplicas,
+	}, nil
+}
+
+func CalculateImportEventGasLimit(
+	ctx context.Context,
+	ethClient ethclient.Client,
+	callData []byte,
+	replica common.Address,
+) (uint64, error) {
+	gasPrice, err := ethClient.SuggestGasPrice(ctx)
+	if err != nil {
+		return 0, err
+	}
+
+	return ethClient.EstimateGas(ctx, interfaces.CallMsg{
+		From:     [20]byte{},
+		To:       &replica,
+		Gas:      0,
+		GasPrice: gasPrice,
+		Value:    big.NewInt(0),
+		Data:     callData,
+	})
+}
+
+func (c *ChainlinkMessageHandler) ShouldSendMessage(destinationClient vms.DestinationClient) (bool, error) {
+	return true, nil
+}
+
+func (c *ChainlinkMessageHandler) SendMessage(
+	signedMessage *warp.Message,
+	destinationClient vms.DestinationClient,
+) (common.Hash, error) {
+	var msg ChainlinkMessage
+	if err := json.Unmarshal(signedMessage.Payload, &msg); err != nil {
+		return common.Hash{}, err
+	}
+	destinationBlockchainID := destinationClient.DestinationBlockchainID()
+	teleporterMessageID, err := teleporterUtils.CalculateMessageID(
+		msg.aggregator,
+		signedMessage.SourceChainID,
+		destinationBlockchainID,
+		msg.roundId,
+	)
+	if err != nil {
+		return common.Hash{}, fmt.Errorf("failed to calculate Teleporter message ID: %w", err)
+	}
+
+	c.logger.Info(
+		"Sending message to destination chain",
+		zap.String("destinationBlockchainID", destinationBlockchainID.String()),
+		zap.String("warpMessageID", signedMessage.ID().String()),
+	)
+
+	callData, err := eventimporter.PackImportEvent(msg.blockHeader, msg.txIndex, msg.receiptProof, msg.logIndex)
+	if err != nil {
+		c.logger.Error(
+			"Failed packing importEvent call data",
+			zap.String("destinationBlockchainID", destinationBlockchainID.String()),
+			zap.String("warpMessageID", signedMessage.ID().String()),
+		)
+		return common.Hash{}, err
+	}
+
+	replica, ok := c.aggregatorsToReplicas[msg.aggregator]
+	if !ok {
+		c.logger.Error(
+			"Failed to find replica for aggregator",
+			zap.String("destinationBlockchainID", destinationBlockchainID.String()),
+			zap.String("warpMessageID", signedMessage.ID().String()),
+			zap.Error(err),
+		)
+		return common.Hash{}, fmt.Errorf("failed to find replica for aggregator: %s", msg.aggregator)
+	}
+	subnetClient := destinationClient.Client().(subnetEthclient.Client)
+	client := reflect.ValueOf(subnetClient).Elem().FieldByName("c").Interface().(ethclient.Client)
```

Comment:
This looks a bit brittle but how can we do a gas estimation given a subnetevm client ?